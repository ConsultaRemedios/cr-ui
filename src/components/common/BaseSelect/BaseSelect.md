### Basic Usage
```js
const options = [
  { label: 'Brasil', value: 1 },
  { label: 'Estados Unidos', value: 2 },
  { label: 'Inglaterra', value: 3 },
  { label: 'China', value: 4 },
  { label: 'Argentina', value: 5 },
  { label: 'Chile', value: 6 },
  { label: 'Nova Zelândia', value: 7 },
  { label: 'França', value: 8 },
  { label: 'Itália', value: 9 },
  { label: 'Dinamarca', value: 10 },
];

<BaseSelect
  :options="options"
  placeholder="Selecione um país"
>
  <BaseOption v-for="option in options"
    :value="option.value"
    :key="option.value"
  >
    {{option.label}}
  </BaseOption>
</BaseSelect>
```

### With Error
```js
const options = [
  { label: 'Brasil', value: 1 },
  { label: 'Estados Unidos', value: 2 },
  { label: 'Inglaterra', value: 3 },
  { label: 'China', value: 4 },
  { label: 'Argentina', value: 5 },
  { label: 'Chile', value: 6 },
  { label: 'Nova Zelândia', value: 7 },
  { label: 'França', value: 8 },
  { label: 'Itália', value: 9 },
  { label: 'Dinamarca', value: 10 },
];

<BaseSelect
  :options="options"
  :has-error="true"
>
  <BaseOption v-for="option in options"
    :value="option.value"
    :key="option.value"
  >
    {{option.label}}
  </BaseOption>
</BaseSelect>
```

### Custom Select
If for some reason you need a custom select, you can do it.
Just provide a value propriety for the option objects and the others props
it's up to you. You just need some steps:

* Create option collection with the required value propriety
* Pass an object with the same option structure as placeholder prop
* Create a template for the selected option
* Pass the custom option template as the children of the option component

```
const options = [
  { title: 'Casa', description: 'Rua alameda dos anjos...', value: 'casa' },
  { title: 'Escritório', description: 'Rua das travessas...', value: 'office' },
];

const style = {
  title: { fontSize: '14px', margin: 0, fontWeight: 600 },
  description: { margin: '0', fontSize: '14px' },
};

<BaseSelect
  :options="options"
  :placeholder="{ title: 'Endereço', description: 'selecione um endereço' }"
  :rounded="false"
>
  <template slot="selected" scope="props">
    <h1 :style="{ ...style.title }">{{props.currentLabel.title}}</h1>
    <p :style="{ ...style.description }">{{props.currentLabel.description}}</p>
  </template>

  <BaseOption v-for="option in options"
    :value="option.value"
    :key="option.value"
  >
    <div>
      <h1 :style="{ ...style.title }">{{option.title}}</h1>
      <p :style="{ ...style.description }">{{option.description}}</p>
    </div>
  </BaseOption>
</BaseSelect>
```

#### With error
```
const options = [
  { title: 'Casa', description: 'Rua alameda dos anjos...', value: 'casa' },
  { title: 'Escritório', description: 'Rua das travessas...', value: 'office' },
];

const style = {
  title: { fontSize: '14px', margin: 0, fontWeight: 600 },
  description: { margin: '0', fontSize: '14px' },
};

<BaseSelect
  :options="options"
  :rounded="false"
  value="office"
  :has-error="true"
>
  <template slot="selected" scope="props">
    <h1 :style="{ ...style.title }">{{props.currentLabel.title}}</h1>
    <p :style="{ ...style.description }">{{props.currentLabel.description}}</p>
  </template>

  <BaseOption v-for="option in options"
    :value="option.value"
    :key="option.value"
  >
    <div>
      <h1 :style="{ ...style.title }">{{option.title}}</h1>
      <p :style="{ ...style.description }">{{option.description}}</p>
    </div>
  </BaseOption>
</BaseSelect>
```
