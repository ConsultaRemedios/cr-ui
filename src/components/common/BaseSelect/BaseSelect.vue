<template>
  <div
    v-click-outside="closeSelect"
    :class="$style.select"
    :style="{ zIndex }"
  >
    <div
      :class="['cr-base-select',
               { ['cr-base-select__has-error']: hasError },
               { ['cr-base-select__rounded']: rounded && !isOpen },
               { ['cr-base-select__top-rounded']: rounded && isOpen },
               { ['cr-base-select__select-opened']: isOpen }
      ]"
    >
      <slot
        name="selected"
        :current-label="currentLabel"
      >
        {{ currentLabel.label }}
      </slot>
    </div>

    <input
      :class="[$style.selectedLabel, {
      }]"
      type="text"
      readonly="readonly"
      :value="currentLabel.value"
      @click="onClickSelect"
      @keydown.prevent.enter="onPressEnter"
      @keydown.prevent.esc="onPressEsc"
      @keydown.prevent.up="navigateUp"
      @keydown.prevent.down="navigateDown"
    >

    <BaseIcon
      :id="icons.nextIcon.id"
      :class="[
        'cr-base-select__svg',
        { ['cr-base-select__svg__error']: hasError },
        { [$style.arrowDown]: !isOpen },
        { [$style.arrowUp]: isOpen }
      ]"
    />

    <div
      v-show="isOpen"
      :class="[
        'cr-base-select__options',
        { ['cr-base-select__options__error']: hasError },
        { ['cr-base-select__options__bottomRounded']: rounded },
        { ['cr-base-select__options__opened']: isOpen }
      ]"
      data-options
    >
      <slot />
    </div>
  </div>
</template>

<script>
import clickOutside from '../../../directives/click-outside';
import BaseIcon from '../BaseIcon';
import nextIcon from '../../../icons/chevron-right.icon.svg';

export default {
  name: 'BaseSelect',

  components: { BaseIcon },

  directives: {
    clickOutside,
  },

  props: {
    placeholder: {
      type: [String, Object],
    },
    options: {
      type: Array,
    },
    name: {
      type: String,
      default: '',
    },
    value: {
      type: [String, Number],
      default: '',
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    rounded: {
      type: Boolean,
      default: true,
    },

    keyValue: {
      type: String,
      default: 'value',
    },
  },

  data() {
    return {
      isOpen: false,
      hovered: null,
      selected: this.value,
    };
  },

  computed: {
    currentLabel() {
      if (this.selected !== '') {
        const selected = this.getOption(this.selected);
        if (selected) return this.getOption(this.selected);

        return {
          label: this.selected,
          value: this.selected,
        };
      }

      return typeof this.placeholder === 'string'
        ? { label: this.placeholder }
        : this.placeholder;
    },

    zIndex() {
      return this.isOpen ? 2 : 1;
    },

    icons() {
      return {
        nextIcon,
      };
    },
  },

  watch: {
    value(value) {
      this.selected = value;
    },
  },

  created() {
    this.$on('change', this.onOptionChange);
    this.$on('hover', this.onOptionHover);

    if (this.value === '' && !this.placeholder) {
      this.$emit('change', this.options[0].value);
    }
  },

  beforeDestroy() {
    this.$off('change');
    this.$off('hover');
  },

  methods: {
    getOption(value) {
      return this.options.find((o) => o[this.keyValue] === value);
    },

    onOptionChange(value) {
      this.selected = value;
      this.closeSelect();
    },

    onOptionHover(hoveredValue) {
      this.hovered = hoveredValue;
    },

    onClickSelect(e) {
      if (this.isOpen) {
        e.currentTarget.blur();
      }

      this.isOpen = !this.isOpen;
    },

    onPressEnter() {
      const value = this.hovered ? this.hovered : this.value;
      this.$emit('change', value);
    },

    onPressEsc(e) {
      this.closeSelect();

      if (e.currentTarget) {
        e.currentTarget.blur();
      }
    },

    closeSelect() {
      this.isOpen = false;
      this.hovered = null;
    },

    navigateUp() {
      this.navigate('prev');
    },

    navigateDown() {
      this.navigate('next');
    },

    navigate(direction) {
      const { length } = this.options;
      let nextOptionIndex = direction === 'next' ? 0 : length - 1;

      if (this.hovered) {
        const option = this.getOption(this.hovered);
        const optionIndex = this.options.indexOf(option);

        nextOptionIndex = direction === 'prev'
          ? optionIndex - 1
          : optionIndex + 1;
      }

      if (nextOptionIndex >= 0 && nextOptionIndex <= length - 1) {
        this.hovered = this.options[nextOptionIndex][this.keyValue];
        this.$nextTick(() => this.scrollToOption(this.hovered));
      }
    },

    scrollToOption(optionValue) {
      const $option = this.$el.querySelector(`[data-option-value="${optionValue}"]`);
      const $options = this.$el.querySelector('[data-options]');

      if ($option.offsetTop < $options.scrollTop) {
        $options.scrollTop = $option.offsetTop;
      } else if ($option.offsetTop + $option.clientHeight > $options.clientHeight) {
        $options.scrollTop = ($option.offsetTop + $option.clientHeight) - $options.clientHeight;
      }
    },
  },
};
</script>

<style module>
  .select {
    cursor: pointer;
    position: relative;
  }

  .selectedBlock {
    font-size: 15px;
    padding: 11px 44px 11px 15px;
    border: 1px solid #DADADA;
    cursor: pointer;
    position: relative;
    z-index: 2;
    background: #FFF;
  }

  .selectedLabel {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    opacity: 0;
    width: 100%;
    height: 100%;
    cursor: pointer;
    z-index: 3;
  }

  .arrow {
    font-size: 20px;
    position: absolute;
    z-index: 2;
    top: 50%;
    right: 20px;
    transform: translateY(-50%);
  }

  .arrowDown {
    transform: translateY(-50%) rotate(90deg);
    composes: arrow;
  }

  .arrowUp {
    composes: arrow;
    transform: translateY(-50%) rotate(-90deg);
  }

  .options {
    padding: 10px 0;
    max-height: 300px;

    position: absolute;
    top: 100%;
    left: 0;
    right: 0;
    transform: translateY(-1px);

    overflow: auto;
    z-index: 1;

    border: 1px solid #DADADA;
    background: #FFF;
    box-shadow: 0 0 10px 0 rgba(0,0,0,0.10);
  }

  .hasError {
    border-color: #D30B3F;
  }

  .rounded {
    border-radius: 3px;
  }

  .topRounded {
    border-radius: 3px 3px 0 0;
  }

  .bottomRounded {
    border-radius: 0 0 3px 3px;
  }
</style>

<style lang="scss">
  .cr-base-select {
    font-size: 15px;
    padding: 11px 44px 11px 15px;
    border: 1px solid #DADADA;
    cursor: pointer;
    position: relative;
    z-index: 2;
    background: #FFF;

    &__rounded {
      border-radius: 3px;
    }

    &__top-rounded {
      border-radius: 3px 3px 0 0;
    }

    &__has-error {
      border-color: #D30B3F;
    }

    &__select-opened {

    }
  }

  .cr-base-select__options {
    padding: 10px 0;
    max-height: 300px;

    position: absolute;
    top: 100%;
    left: 0;
    right: 0;
    transform: translateY(-1px);

    overflow: auto;
    z-index: 1;

    border: 1px solid #DADADA;
    background: #FFF;
    box-shadow: 0 0 10px 0 rgba(0,0,0,0.10);

    &__bottom-rounded {
      border-radius: 0 0 3px 3px;
    }
  }
</style>
