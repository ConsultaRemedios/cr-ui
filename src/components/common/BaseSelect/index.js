import BaseSelect from './BaseSelect.vue';

export default BaseSelect;
