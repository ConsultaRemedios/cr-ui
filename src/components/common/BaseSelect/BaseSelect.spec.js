import { shallowMount } from '@vue/test-utils';
import BaseSelect from './BaseSelect.vue';

const setup = (props = {}) => {
  const options = [{
    label: 'Option 2',
    value: 1,
  }, {
    label: 'Option 1',
    value: 0,
  }];

  const wrapper = shallowMount(BaseSelect, {
    propsData: {
      value: 0,
      options,
      ...props,
    }
  });

  return { wrapper };
};

describe('BaseSelect', () => {
  describe('matches the snapshot', () => {
    let wrapper;
    beforeEach(() => {
      ({ wrapper } = setup());
    });

    it('when component is mounted', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  it('updates selected data when value prop changes', () => {
    const { wrapper } = setup();

    wrapper.setProps({ value: 1 });

    expect(wrapper.vm.selected).toEqual(1);

    wrapper.setProps({ value: 0 });

    expect(wrapper.vm.selected).toEqual(0);
  });

  describe('on create', () => {
    describe('With no placeholder', () => {
      it('does not set the first option as selected when the value is 0', () => {
        const { wrapper } = setup();
        expect(wrapper.vm.currentLabel).toEqual({
          label: 'Option 1',
          value: 0,
        });
      });

      it('sets the first option as selected when the value is ""', () => {
        const { wrapper } = setup({ value: '' });
        expect(wrapper.vm.currentLabel).toEqual({
          label: 'Option 2',
          value: 1,
        });
      });
    });
  });

  describe('computed props', () => {
    describe('#currentLabel', () => {
      it('does not set placeholder as currentLabel when selected value is 0', () => {
        const { wrapper } = setup({ placeholder: 'Selecione' });
        expect(wrapper.vm.currentLabel).toEqual({
          label: 'Option 1',
          value: 0,
        });
      });

      it('set placeholder as currentLabel when selected value is ""', () => {
        const { wrapper } = setup({ value: '', placeholder: 'Selecione' });
        expect(wrapper.vm.currentLabel).toEqual({
          label: 'Selecione'
        });
      });

      it('set currentLabel when option selet not present in options', () => {
        const { wrapper } = setup({ value: '', placeholder: 'Selecione' });
        wrapper.setData({ selected: 3 });
        expect(wrapper.vm.currentLabel).toEqual({
          label: 3,
          value: 3,
        });
      });
    });
  });
});
