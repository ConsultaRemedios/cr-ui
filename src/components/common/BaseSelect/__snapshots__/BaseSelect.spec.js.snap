// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseSelect matches the snapshot #styles when "listOptionsClass" is passed 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -20,10 +20,10 @@
      class="arrowDown"
      src="./../../../assets/select-arrow.svg"
    />
     
    <div
-     class="options bottomRounded"
+     class="options bottomRounded $style.listOptionsClassStyle"
      data-options=""
      style="display: none;"
    />
  </div>
`;

exports[`BaseSelect matches the snapshot #styles when "selectedBlockClass" is passed 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -1,11 +1,11 @@
  <div
    class="select"
    style="z-index: 1;"
  >
    <div
-     class="selectedBlock rounded"
+     class="selectedBlock $style.selectedBlockStyle rounded"
    >
      
        Option 1
      
    </div>
`;

exports[`BaseSelect matches the snapshot #styles when "selectedBlockErrorClass" is passed 1`] = `
Snapshot Diff:
Compared values have no visual difference.
`;

exports[`BaseSelect matches the snapshot when component is mounted 1`] = `
<div
  class="select"
  style="z-index: 1;"
>
  <div
    class="selectedBlock rounded"
  >
    
      Option 1
    
  </div>
   
  <input
    class="selectedLabel"
    readonly="readonly"
    type="text"
  />
   
  <img
    class="arrowDown"
    src="./../../../assets/select-arrow.svg"
  />
   
  <div
    class="options bottomRounded"
    data-options=""
    style="display: none;"
  />
</div>
`;
